pkg/apis/migration/migration.go

```diff
@@ -92,46 +92,51 @@ const (
 	CRDPlural = "cnsvspherevolumemigrations"
 )
 
-// onceForVolumeMigrationService is used for initializing the VolumeMigrationService singleton.
-var onceForVolumeMigrationService sync.Once
-
-// volumeMigrationInstance is instance of volumeMigration and implements interface for VolumeMigrationService
-var volumeMigrationInstance *volumeMigration
-
-var volumeMigrationServiceInitErr error
+var (
+	// volumeMigrationInstance is instance of volumeMigration and implements interface for VolumeMigrationService
+	volumeMigrationInstance *volumeMigration
+	// volumeMigrationInstanceLock is used for handling race conditions during read, write on volumeMigrationInstance
+	volumeMigrationInstanceLock = &sync.RWMutex{}
+)
 
 // GetVolumeMigrationService returns the singleton VolumeMigrationService
 func GetVolumeMigrationService(ctx context.Context, volumeManager *cnsvolume.Manager, cnsConfig *cnsconfig.Config) (VolumeMigrationService, error) {
 	log := logger.GetLogger(ctx)
-	onceForVolumeMigrationService.Do(func() {
+	volumeMigrationInstanceLock.RLock()
+	if volumeMigrationInstance == nil {
+		volumeMigrationInstanceLock.RUnlock()
+		volumeMigrationInstanceLock.Lock()
+		defer volumeMigrationInstanceLock.Unlock()
 		log.Info("Initializing volume migration service...")
 		volumeMigrationInstance = &volumeMigration{
```

Comment:
Shouldn't you be checking one more time if volumeMigrationInstance is initialized or not? Otherwise, multiple threads may be initializing different instance after getting the write lock.

Reply:
My bad.. I think I missed it in an initial commit. I have squashed all commits for now. Please take a look